File: libs/noug/event.ts
/* ***** BEGIN LICENSE BLOCK *****
 * Distributed under the BSD license:
 *
 * Copyright (c) 2015, xuewen.chu
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of xuewen.chu nor the
 *       names of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL xuewen.chu BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * ***** END LICENSE BLOCK ***** */

import {Event,EventNoticer,Notification} from './_event';
export * from './_event';
import {Action} from './_action';
import {View} from './_view';

export default (exports.event as (target: any, name: string)=>void);

// GUI EVENT 

export enum ClickType {
	TOUCH = 1, KEYBOARD = 2, MOUSE = 3,
};

export enum HighlightedStatus {
	HIGHLIGHTED_NORMAL = 1,
	HIGHLIGHTED_HOVER,
	HIGHLIGHTED_DOWN,
};

export enum ReturnValueMask {
	RETURN_VALUE_MASK_DEFAULT = (1 << 0),
	RETURN_VALUE_MASK_BUBBLE = (1 << 1),
	RETURN_VALUE_MASK_ALL = (RETURN_VALUE_MASK_DEFAULT | RETURN_VALUE_MASK_BUBBLE),
};

export enum KeyboardKeyName {
	UNKNOWN =          0,
	MOUSE_LEFT =       1,
	MOUSE_CENTER =     2,
	MOUSE_RIGHT =      3,
	MOUSE_WHEEL_UP =   4,
	MOUSE_WHEEL_DOWN = 5,
	MOUSE_WHEEL_LEFT = 6,
	MOUSE_WHEEL_RIGHT =7,
	BACK_SPACE =       8,  /* back space */
	TAB =              9,  /* tab */
	CLEAR =            12, /* clear */
	ENTER =            13, /* enter */
	SHIFT =            16, /* shift */
	CTRL =             17, /* ctrl */
	ALT =              18, /* alt */
	CAPS_LOCK =        20, /* caps lock */
	ESC =              27, /* esc */
	SPACE =            32, /* space */
	COMMAND =          91, /* command/win */
	LEFT =             37, /* left */
	UP =               38, /* up */
	RIGHT =            39, /* right */
	DOWN =             40, /* down */
	INSERT =           45, /* insert */
	DELETE =           46, /* delete */
	PAGE_UP =          33, /* page up */
	PAGE_DOWN =        34, /* page down */
	MOVE_END =         35, /* end */
	MOVE_HOME =        36, /* home */
	SCROLL_LOCK =      145,
	BREAK =            19,
	SYSRQ =            42,
	HELP =             47, /* Help */
	NUM_0 =            48, /* 0 ) */
	NUM_1 =            49, /* 1 ! */
	NUM_2 =            50, /* 2 @ */
	NUM_3 =            51, /* 3 # */
	NUM_4 =            52, /* 4 $ */
	NUM_5 =            53, /* 5 % */
	NUM_6 =            54, /* 6 ^ */
	NUM_7 =            55, /* 7 & */
	NUM_8 =            56, /* 8 * */
	NUM_9 =            57, /* 9 ( */
	A =                65, /* a-z */
	B =                66,
	C =                67,
	D =                68,
	E =                69,
	F =                70,
	G =                71,
	H =                72,
	I =                73,
	J =                74,
	K =                75,
	L =                76,
	M =                77,
	N =                78,
	O =                79,
	P =                80,
	Q =                81,
	R =                82,
	S =                83,
	T =                84,
	U =                85,
	V =                86,
	W =                87,
	X =                88,
	Y =                89,
	Z =                90,
	NUM_LOCK =         144, /* numpad */
	NUMPAD_0 =         96,
	NUMPAD_1 =         97,
	NUMPAD_2 =         98,
	NUMPAD_3 =         99,
	NUMPAD_4 =         100,
	NUMPAD_5 =         101,
	NUMPAD_6 =         102,
	NUMPAD_7 =         103,
	NUMPAD_8 =         104,
	NUMPAD_9 =         105,
	NUMPAD_DIVIDE =    111, /* / */
	NUMPAD_MULTIPLY =  106,  /* * */
	NUMPAD_SUBTRACT =  109,  /* - */
	NUMPAD_ADD =       107, /* + */
	NUMPAD_DOT =       110, /* . */
	NUMPAD_ENTER =     108, /* enter */
	F1 =               112, /* f1 - f24 */
	F2 =               113,
	F3 =               114,
	F4 =               115,
	F5 =               116,
	F6 =               117,
	F7 =               118,
	F8 =               119,
	F9 =               120,
	F10 =              121,
	F11 =              122,
	F12 =              123,
	F13 =              124,
	F14 =              125,
	F15 =              126,
	F16 =              127,
	F17 =              128,
	F18 =              129,
	F19 =              130,
	F20 =              131,
	F21 =              132,
	F22 =              133,
	F23 =              134,
	F24 =              135,
	SEMICOLON =        186,  /* ; : */
	EQUALS =           187,  /* = + */
	MINUS =            189,  /* - _ */
	COMMA =            188,  /* , < */
	PERIOD =           190,  /* . > */
	SLASH =            191,  /* / ? */
	GRAVE =            192,  /* ` ~ */
	LEFT_BRACKET =     219,  /* [ { */
	BACK_SLASH =       220,  /* \ | */
	RIGHT_BRACKET =    221,  /* ] } */
	APOSTROPHE =       222,  /* ' " */
	/* --------------------------------------------------- */
	HOME =             300,     /* 按键Home */
	BACK =             301,     /* 返回键 */
	CALL =             302,     /* 拨号键 */
	ENDCALL =          303,     /* 挂机键 */
	STAR =             304,     /* * */
	POUND =            305,     /* # */
	CENTER =           306,     /* 导航键 确定键 */
	VOLUME_UP =        307,     /* 音量增加键 */
	VOLUME_DOWN =      308,     /* 音量减小键 */
	POWER =            309,     /* 电源键 */
	CAMERA =           310,     /* 拍照键 */
	FOCUS =            311,     /* 拍照对焦键 */
	MENU =             312,     /* 菜单键 */
	SEARCH =           313,     /* 搜索键 */
	MEDIA_PLAY_PAUSE = 314,     /* 多媒体键 播放/暂停 */
	MEDIA_STOP =       315,     /* 多媒体键 停止 */
	MEDIA_NEXT =       316,     /* 多媒体键 下一首 */
	MEDIA_PREVIOUS =   317,     /* 多媒体键 上一首 */
	MEDIA_REWIND =     318,     /* 多媒体键 快退 */
	MEDIA_FAST_FORWARD =319,    /* 多媒体键 快进 */
	MUTE =             320,     /* 话筒静音键 */
	CHANNEL_UP =       321,     /* 按键Channel up */
	CHANNEL_DOWN =     322,     /* 按键Channel down */
	MEDIA_PLAY =       323,     /* 多媒体键 播放 */
	MEDIA_PAUSE =      324,     /* 多媒体键 暂停 */
	MEDIA_CLOSE =      325,     /* 多媒体键 关闭 */
	MEDIA_EJECT =      326,     /* 多媒体键 弹出 */
	MEDIA_RECORD =     327,     /* 多媒体键 录音 */
	VOLUME_MUTE =      328,     /* 扬声器静音键 */
	MUSIC =            329,     /* music */
	EXPLORER =         330,     /* 按键Explorer special function */
	ENVELOPE =         331,     /* 按键Envelope special function */
	BOOKMARK =         332,     /* 按键Bookmark */
	ZOOM_IN =          333,     /* 放大键 */
	ZOOM_OUT =         334,     /* 缩小键 */
}

declare class NativeEvent<Sender, Data, Origin> extends Event<Sender, Data, Origin> {}

export declare class UIEvent<Data = any> extends NativeEvent<View, Data, View> {
	readonly timestamp: number;
	readonly isDefault: boolean;
	readonly isBubble: boolean;
	cancelDefault(): void;
	cancelBubble(): void;
}

export declare class ActionEvent<Data = any> extends UIEvent<Data> {
	readonly action: Action;
	readonly delay: number;
	readonly frame: number;
	readonly loop: number;
}

export declare class KeyEvent<Data = any> extends UIEvent<Data> {
	readonly keycode: number;
	readonly repeat: number;
	readonly shift: boolean;
	readonly ctrl: boolean;
	readonly alt: boolean;
	readonly command: boolean;
	readonly capsLock: boolean;
	readonly device: number;
	readonly source: number;
	focusMove: View | null;
}

export declare class ClickEvent<Data = any> extends UIEvent<Data> {
	readonly x: number;
	readonly y: number;
	readonly count: number;
	readonly type: ClickType;
}

export declare class HighlightedEvent<Data = any> extends UIEvent<Data> {
	readonly status: HighlightedStatus;
}

export declare class MouseEvent<Data = any> extends UIEvent<Data> {
	readonly x: number;
	readonly y: number;
}

export interface TouchPoint {
	readonly id: number;
	readonly startX: number;
	readonly startY: number;
	readonly x: number; 
	readonly y: number;
	readonly force: number;
	readonly clickIn: boolean;
	readonly view: View;
}

export declare class TouchEvent<Data = any> extends UIEvent<Data> {
	readonly changedTouches: TouchPoint[];
}

export declare class FocusMoveEvent<Data = any> extends UIEvent<Data> {
	readonly focus: View | null;
	readonly focusMove: View | null;
}

const _util = __require__('_util');
const PREFIX = 'm_on';

/**
 * @class NativeNotification
 */
export class NativeNotification<E = Event> extends Notification<E> {

	getNoticer(name: string) {
		var noticer = (this as any)[PREFIX + name] as EventNoticer<E>;
		if ( ! noticer ) {
			// bind native event
			var func = (this as any)['trigger' + name];
			// bind native
			_util.addNativeEventListener(this, name, (event?: any, isEvent?: boolean) => {
				//console.log('_util.addNativeEventListener', name);
				var evt = event && isEvent ? event: new Event(event);
				var ok = func ? func.call(this, evt): this.triggerWithEvent(name, evt);
				//console.log('_util.addNativeEventListener', name, ok, String(trigger));
				return ok;
			}, -1);
			(this as any)[PREFIX + name] = noticer = new EventNoticer<E>(name, this as any);
		}
		return noticer;
	}
	
}

File: src/js/api/_/event.cc
/* ***** BEGIN LICENSE BLOCK *****
 * Distributed under the BSD license:
 *
 * Copyright (c) 2015, blue.chu
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of blue.chu nor the
 *       names of its contributors may be used to endorse or promote products
 *       derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL blue.chu BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
 * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * ***** END LICENSE BLOCK ***** */

#include "../types.h"
#include "./view.h"
#include "../../ui/event.h"
#include "../../ui/action/action.h"
#include "../../ui/view/button.h"
#include "../../ui/keyboard.h"
#include <native-inl-js.h>

namespace qk { namespace js {
	typedef Event<> NativeEvent;

	class WrapNativeEvent: public WrapObject {
	public:
		typedef Event<> Type;
		
		static void constructor(FunctionArgs args) {
			Js_ATTACH(args);
			Js_Worker(args);
			Js_Throw("Access forbidden.");
		}
		
		static void noticer(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_UNPACK(Type);
			Js_Return( wrap->get( worker->strs()->_noticer() ) );
		}
		
		static void sender(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_UNPACK(Type);
			JSValue* noticer = wrap->get(worker->strs()->_noticer());
			
			if ( !noticer.IsEmpty() && noticer->IsObject(worker) ) {
				JSValue* sender = noticer.To<JSObject>()->Get(worker, worker->strs()->sender());
				if ( !sender.IsEmpty() ) {
					Js_Return(sender);
				}
			} else {
				if ( wrap->self()->noticer() ) {
					Wrap<Object>* sender = Wrap<Object>::pack( wrap->self()->sender() );
					if ( sender ) {
						Js_Return( sender->that() );
					}
				}
			}
			Js_Return_Null();
		}
		
		static void name(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_UNPACK(Type);
			JSValue* noticer = wrap->get(worker->strs()->_noticer());
			
			if ( !noticer.IsEmpty() && noticer->IsObject(worker) ) {
				JSValue* name = noticer.To<JSObject>()->Get(worker, worker->strs()->name());
				if ( !name.IsEmpty() ) {
					Js_Return(name);
				}
			} else {
				if ( wrap->self()->noticer() ) {
					Js_Return( wrap->self()->name() );
				} else {
					Js_Return_Null();
				}
			}
		}
		
		static void data(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_UNPACK(Type);
			Cast* cast = static_cast<Cast*>(wrap->privateData());
			if ( cast ) {
				Js_Return( cast->cast(*wrap->self()->data(), worker) );
			}
			Js_Return_Null();
		}

		static void origin(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Return_Null();
		}
		
		static void return_value(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(Type);
			Js_Return( self->return_value );
		}
		
		static void set_return_value(JSString* name,
																JSValue* value, PropertySetArgs args) {
			Js_Worker(args);
			if ( !value->IsBoolean(worker) && !value->IsNumber(worker) ) {
				Js_Throw("Bad argument.");
			}
			Js_Self(Type);
			self->return_value = value->ToInt32Value(worker);
		}
		
	public:
		static void binding(JSObject* exports, Worker* worker) {
			Js_NEW_CLASS(NativeEvent, constructor, {
				Js_Set_Class_Accessor(noticer, noticer);
				Js_Set_Class_Accessor(sender, sender);
				Js_Set_Class_Accessor(data, data);
				Js_Set_Class_Accessor(name, name);
				Js_Set_Class_Accessor(origin, origin);
				Js_Set_Class_Accessor(returnValue, return_value, set_return_value);
			}, exports->GetProperty(worker, "Event").To<JSFunction>());
			
			cls->Export(worker, "NativeEvent", exports);
		}
	};

	class WrapUIEvent: public WrapObject {
		public:
		typedef UIEvent Type;
		
		static void constructor(FunctionArgs args) {
			Js_ATTACH(args);
			Js_Worker(args);
			Js_Throw("Access forbidden.");
		}

		static void origin(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(Type);
			if ( self->origin() ) {
				Wrap<View>* wrap = Wrap<View>::pack(self->origin());
				Js_Return( wrap->that() );
			} else {
				Js_Return_Null();
			}
		}

		static void timestamp(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(Type);
			Js_Return( self->timestamp() );
		}

		static void cancel_default(FunctionArgs args) {
			Js_Worker(args);
			Js_Self(Type);
			self->cancel_default();
		}

		static void cancel_bubble(FunctionArgs args) {
			Js_Worker(args);
			Js_Self(Type);
			self->cancel_bubble();
		}

		static void is_default(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(Type);
			Js_Return( self->is_default() );
		}

		static void is_bubble(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(Type);
			Js_Return( self->is_bubble() );
		}
		
	public:
		static void binding(JSObject* exports, Worker* worker) {
			Js_Define_Class(UIEvent, constructor, {
				Js_Set_Class_Accessor(origin, origin);
				Js_Set_Class_Accessor(timestamp, timestamp);
				Js_Set_Class_Method(cancelDefault, cancel_default);
				Js_Set_Class_Method(cancelBubble, cancel_bubble);
				Js_Set_Class_Accessor(isDefault, is_default);
				Js_Set_Class_Accessor(isBubble, is_bubble);
			}, Event<>);
		}
	};

	class WrapUIActionEvent: public WrapObject {
		public:
		typedef UIActionEvent Type;
		
		static void constructor(FunctionArgs args) {
			Js_ATTACH(args);
			Js_Worker(args);
			Js_Throw("Access forbidden.");
		}

		static void action(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(UIActionEvent);
			if ( self->action() ) {
				Wrap<Action>* wrap = Wrap<Action>::pack(self->action());
				Js_Return( wrap->that() );
			} else {
				Js_Return_Null();
			}
		}

		static void delay(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(UIActionEvent);
			Js_Return( self->delay() / 1000 );
		}

		static void frame(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(UIActionEvent);
			Js_Return( self->frame() );
		}

		static void loop(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(UIActionEvent);
			Js_Return( self->loop() );
		}
		
		static void binding(JSObject* exports, Worker* worker) {
			Js_Define_Class(UIActionEvent, constructor, {
				Js_Set_Class_Accessor(action, action);
				Js_Set_Class_Accessor(delay, delay);
				Js_Set_Class_Accessor(frame, frame);
				Js_Set_Class_Accessor(loop, loop);
			}, UIEvent);
		}
	};

	class WrapUIKeyEvent: public WrapObject {
		public:
		typedef UIKeyEvent Type;
		
		static void constructor(FunctionArgs args) {
			Js_ATTACH(args);
			Js_Worker(args);
			Js_Throw("Access forbidden.");
		}

		static void keycode(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(Type);
			Js_Return( self->keycode() );
		}

		static void repeat(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(Type);
			Js_Return( self->repeat() );
		}

		static void shift(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(Type);
			Js_Return( self->shift() );
		}

		static void ctrl(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(Type);
			Js_Return( self->ctrl() );
		}

		static void alt(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(Type);
			Js_Return( self->alt() );
		}

		static void command(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(Type);
			Js_Return( self->command() );
		}

		static void caps_lock(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(Type);
			Js_Return( self->caps_lock() );
		}

		static void device(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(Type);
			Js_Return( self->device() );
		}

		static void source(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(Type);
			Js_Return( self->source() );
		}

		static void focus_move(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(Type);
			if ( self->focus_move() ) {
				Js_Return( Wrap<View>::pack(self->focus_move(), View::VIEW)->that() );
			} else {
				Js_Return_Null();
			}
		}

		static void set_focus_move(JSString* name, JSValue* value, PropertySetArgs args) {
			Js_Worker(args);
			Js_Self(Type);
			View* view = nullptr;
			if ( worker->hasInstance(value, View::VIEW) ) {
				view = WrapObject::unpack<Button>(value.To<JSObject>())->self();
			} else if ( ! value->IsNull(worker) ) {
				Js_Throw("Bad argument.");
			}
			self->set_focus_move(view);
		}
		
	public:
		static void binding(JSObject* exports, Worker* worker) {
			Js_Define_Class(UIKeyEvent, constructor, {
				Js_Set_Class_Accessor(keycode, keycode);
				Js_Set_Class_Accessor(repeat, repeat);
				Js_Set_Class_Accessor(shift, shift);
				Js_Set_Class_Accessor(ctrl, ctrl);
				Js_Set_Class_Accessor(alt, alt);
				Js_Set_Class_Accessor(command, command);
				Js_Set_Class_Accessor(capsLock, caps_lock);
				Js_Set_Class_Accessor(device, device);
				Js_Set_Class_Accessor(source, source);
				Js_Set_Class_Accessor(focusMove, focus_move, set_focus_move);
			}, UIEvent);
		}
	};

	class WrapUIClickEvent: public WrapObject {
		public:
		typedef UIClickEvent Type;
		
		static void constructor(FunctionArgs args) {
			Js_ATTACH(args);
			Js_Worker(args);
			Js_Throw("Access forbidden.");
		}

		static void x(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(UIClickEvent);
			Js_Return( self->x() );
		}

		static void y(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(UIClickEvent);
			Js_Return( self->y() );
		}

		static void count(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(UIClickEvent);
			Js_Return( self->count() );
		}

		static void type(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(UIClickEvent);
			Js_Return( int(self->type()) );
		}
		
		static void binding(JSObject* exports, Worker* worker) {
			Js_Define_Class(UIClickEvent, constructor, {
				Js_Set_Class_Accessor(x, x);
				Js_Set_Class_Accessor(y, y);
				Js_Set_Class_Accessor(type, type);
				Js_Set_Class_Accessor(count, count);
			}, UIEvent);
		}
	};

	class WrapUIHighlightedEvent: public WrapObject {
		public:
		
		static void constructor(FunctionArgs args) {
			Js_ATTACH(args);
			Js_Worker(args);
			Js_Throw("Access forbidden.");
		}
		
		/**
		 * @get status {HighlightedStatus}
		 */
		static void status(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(UIHighlightedEvent);
			Js_Return( self->status() );
		}
		
		static void binding(JSObject* exports, Worker* worker) {
			Js_Define_Class(UIHighlightedEvent, constructor, {
				Js_Set_Class_Accessor(status, status);
			}, UIEvent);
		}
	};

	class WrapUIMouseEvent: public WrapObject {
		public:
		typedef UIMouseEvent Type;
		
		static void constructor(FunctionArgs args) {
			Js_ATTACH(args);
			Js_Worker(args);
			Js_Throw("Access forbidden.");
		}
		
		static void x(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(UIMouseEvent);
			Js_Return( self->x() );
		}
		
		static void y(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(UIMouseEvent);
			Js_Return( self->y() );
		}
		
	public:
		static void binding(JSObject* exports, Worker* worker) {
			Js_Define_Class(UIMouseEvent, constructor, {
				Js_Set_Class_Accessor(x, x);
				Js_Set_Class_Accessor(y, y);
			}, UIKeyEvent);
		}
	};

	class WrapUITouchEvent: public WrapObject {
		public:
		typedef UITouchEvent Type;
		
		static void constructor(FunctionArgs args) {
			Js_ATTACH(args);
			Js_Worker(args);
			Js_Throw("Access forbidden.");
		}

		static void changedTouches(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_UNPACK(UITouchEvent);
			Js_Handle_Scope();
			
			JSValue* r = wrap->get(worker->strs()->_change_touches());
			
			if (r.IsEmpty()) { // js error
				return;
			}
			
			if ( !r->IsArray(worker) ) {
				Local<JSArray> arr = worker->NewArray();
				int j = 0;
				
				for ( auto& i : wrap->self()->changed_touches() ) {
					JSObject* item = worker->NewObject();
					//
					Wrap<View>* view = Wrap<View>::pack(i.value().view);
					item->Set(worker,worker->strs()->id(), worker->New(i.value().id));
					item->Set(worker,worker->strs()->startX(), worker->New(i.value().start_x));
					item->Set(worker,worker->strs()->startY(), worker->New(i.value().start_y));
					item->Set(worker,worker->strs()->x(), worker->New(i.value().x));
					item->Set(worker,worker->strs()->y(), worker->New(i.value().y));
					item->Set(worker,worker->strs()->force(), worker->New(i.value().force));
					item->Set(worker,worker->strs()->clickIn(), worker->New(i.value().click_in));
					item->Set(worker,worker->strs()->view(), view->that());
					//
					arr->Set(worker, j, item);
					j++;
				}
				
				r = arr.To<JSValue>();
				
				wrap->set(worker->strs()->_change_touches(), r);
			}
			Js_Return( r );
		}
		
	public:
		static void binding(JSObject* exports, Worker* worker) {
			Js_Define_Class(UITouchEvent, constructor, {
				Js_Set_Class_Accessor(changedTouches, changedTouches);
			}, UIEvent);
		}
	};

	class WrapUIFocusMoveEvent: public WrapObject {
		public:
		typedef UIFocusMoveEvent Type;
		
		static void constructor(FunctionArgs args) {
			Js_ATTACH(args);
			Js_Worker(args);
			Js_Throw("Access forbidden.");
		}

		static void focus(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(UIFocusMoveEvent);
			
			if ( self->focus() ) {
				Js_Return( Wrap<View>::pack(self->focus(), View::VIEW)->that() );
			} else {
				Js_Return_Null();
			}
		}

		static void focusMove(JSString* name, PropertyArgs args) {
			Js_Worker(args);
			Js_Self(UIFocusMoveEvent);
			
			if ( self->focus_move() ) {
				Js_Return( Wrap<View>::pack(self->focus_move(), View::VIEW)->that() );
			} else {
				Js_Return_Null();
			}
		}
		
	public:
		static void binding(JSObject* exports, Worker* worker) {
			Js_Define_Class(UIFocusMoveEvent, constructor, {
				Js_Set_Class_Accessor(focus, focus);
				Js_Set_Class_Accessor(focusMove, focusMove);
			}, UIEvent);
		}
	};

	class BindingNativeEvent {
	public:
		static void binding(JSObject* exports, Worker* worker) {
			worker->runNativeScript(WeakBuffer((Char*)
															native_js::INL_native_js_code__event_,
															native_js::INL_native_js_code__event_count_).buffer(), "_event.js", exports);
			WrapNativeEvent::binding(exports, worker);
			WrapUIEvent::binding(exports, worker);
			WrapUIActionEvent::binding(exports, worker);
			WrapUIKeyEvent::binding(exports, worker);
			WrapUIClickEvent::binding(exports, worker);
			WrapUIMouseEvent::binding(exports, worker);
			WrapUITouchEvent::binding(exports, worker);
			WrapUIFocusMoveEvent::binding(exports, worker);
			WrapUIHighlightedEvent::binding(exports, worker);
		}
	};

	Js_Set_Module(_event, BindingNativeEvent)
} }